examples/mario/src/main/scala/Effects.scala

```diff
@@ -0,0 +1,90 @@
+package mario
+
+import org.scalajs.dom
+import org.scalajs.dom.{KeyboardEvent, document}
+import org.scalajs.dom.raw.{Event, HTMLAudioElement, TouchEvent}
+import scalm.{Cmd, Sub, Task}
+import scalm.Sub.ofTotalObservable
+
+object Effects {
+
+  val requestAnimationFrameSub: Sub[Double] = ofTotalObservable[Double](
+    "requestAnimation", { observer =>
+      var handle = 0
+      def loop(time: Double): Unit = {
+        observer.onNext(time)
+        handle = dom.window.requestAnimationFrame(loop)
+      }
+      handle = dom.window.requestAnimationFrame(loop)
+      () =>
+        dom.window.cancelAnimationFrame(handle)
+    }
+  )
+
+  def keyPressSub[M](keyCode: Int, msg: M): Sub[M] = ofTotalObservable[M](
+    s"keyDown$keyCode", { observer =>
+      val listener = (keyEvent: KeyboardEvent) => {
+        if (keyEvent.keyCode == keyCode) observer.onNext(msg)
+      }
+      dom.window.addEventListener("keydown", listener)
+      () =>
+        dom.window.removeEventListener("keydown", listener)
+    }
+  )
+
+  def keyReleaseSub[M](keyCode: Int, msg: M): Sub[M] = ofTotalObservable[M](
+    s"keyUp$keyCode", { observer =>
+      val listener = (keyEvent: KeyboardEvent) => {
+        if (keyEvent.keyCode == keyCode) observer.onNext(msg)
+      }
+      dom.window.addEventListener("keyup", listener)
+      () =>
+        dom.window.removeEventListener("keyup", listener)
+    }
+  )
+
+  val touchStartSub: Sub[(Double, Double)] =
+    ofTotalObservable[(Double, Double)](
+      s"touchStart", { observer =>
+        val listener = (touchEvent: TouchEvent) => {
+          observer.onNext(
+            (touchEvent.touches.item(0).clientX,
+             touchEvent.touches.item(0).clientY))
+        }
+        dom.window.addEventListener("touchstart", listener)
+        () =>
+          dom.window.removeEventListener("touchstart", listener)
+      }
+    )
+
+  val touchEndSub: Sub[(Double, Double)] =
+    ofTotalObservable[(Double, Double)](
+      s"touchEnd", { observer =>
+        val listener = (touchEvent: TouchEvent) => {
+          observer.onNext(
+            (touchEvent.changedTouches.item(0).clientX,
+             touchEvent.changedTouches.item(0).clientY))
+        }
+        dom.window.addEventListener("touchend", listener)
+        () =>
+          dom.window.removeEventListener("touchend", listener)
+      }
+    )
```

Comment:
What do you think of adding the following to scalm core?

~~~ scala
def windowEvent[A, B](name: String, extract: A => Option[B]): Sub[B] =
  Sub.ofTotalObservable[A](name, { observer =>
    val listener = fun { (a: A) =>
      extract(a) match {
        case Some(b) => observer.onNext(b)
        case None => ()
      }
    }
    dow.window.addEventListener(name, listener)
    () => dom.window.removeEventListener(name, listener)
  })

def windowEvent[A](name: String): Sub[A] = globalEvent[A, A](name, a => Some(a))
~~~

Then you could define `touchEndSub` as follows:

~~~ scala
val touchEndSub: Sub[(Double, Double)] =
  windowEvent[TouchEvent]("touchEnd")
    .map(e => (e.changedTouches.item(0).clientX, e.changedTouches.item(0).clientY))
~~~

And `keyPressSub` could also be defined as:

~~~ scala
def keyPressSub[M](keyCode: Int, msg: M): Sub[M] =
  windowEvent(s"keyDown$keyCode", { (evt: KeyboardEvent) =>
    if (evt.keyCode == keyCode) Some(msg) else None
  })
~~~

(I’m not sure the overloading would play well, but we could find distinct names if that’s an issue)

Reply:
It is a good idea indeed. 
We could make it even more generic with something like: 
`def fromEvent[A, B](name: String, target: EventTarget)(extract: A => Option[B]): Sub[B]` where `target: EventTarget` would be of type 'org.scalajs.dom.raw.EventTarget'.
I think having this it can be used with (I guess) any, not just element but even WebSocket or any emitable object.
Use cases:
```
def fromEvent[A, B](name: String, target: EventTarget)(extract: A => Option[B]): Sub[B] = ???

  val ws = new WebSocket()
  val wsSub = fromEvent[MessageEvent, String]("message",ws){ msg => Some(msg.data.toString) }

  val touchendSub = fromEvent[TouchEvent, Double]("touchend", dom.window){ event => Some(event.touches.item(0).clientX) }
```

Reply:
I’m skeptical for the websocket example: IMO the websocket creation should be handled by the subscription itself, not outside of the subscription.